### PR TITLE
[LogCollector] Fix get_log size calculation

### PR DESCRIPTION
**Pull request by liranbg**
seems like the size and offset calculation were a bit off wrt calculating the chunk size to read
add some UT to cover its usages

https://jira.iguazeng.com/browse/ML-4747

### CODE REVIEW

File: server/log-collector/pkg/services/logcollector/server.go [hunk 1]
```diff
@@ -1055,22 +1066,29 @@ func (s *Server) isLogCollectionRunning(ctx context.Context, runUID, project str
...
-	// if the remaining size is smaller than the buffer size, use the remaining size
-	if endSize-currentOffset < chunkSize {
-		chunkSize = endSize - currentOffset
+	// if we didn't read anything yet, assume read the initial offset
+	if totalLogsSize == 0 {
+		totalLogsSize += initialOffset
 	}
 
-	return chunkSize
+	// if the size we need to read is bigger that buffer, use the buffer size
```
Comment (TomerShor) [inline server/log-collector/pkg/services/logcollector/server.go:1085]:
```suggestion
	// if the size we need to read is bigger than the buffer, use the buffer size
```

File: server/log-collector/pkg/services/logcollector/server.go [hunk 2]
```diff
@@ -372,14 +367,27 @@ func (s *Server) GetLogs(request *protologcollector.GetLogsRequest, responseStre
 		return nil
 	}
 
+	var endSize = currentLogFileSize
+
+	// if request size is positive, read only the requested size, add the offset to ensure we read
+	// all requested logs
+	if request.Size > 0 {
+		endSize = request.Size + request.Offset
+	}
```
Comment (TomerShor) [inline server/log-collector/pkg/services/logcollector/server.go:376]: not sure I understand why you add the offset here to the size - we use the offset in `readAt` later on, so the size of the read logs doesn't include the offset bytes...

File: server/log-collector/pkg/services/logcollector/server.go [hunk 3]
```diff
@@ -1055,22 +1066,29 @@ func (s *Server) isLogCollectionRunning(ctx context.Context, runUID, project str
...
+	if totalLogsSize+initialOffset >= endSize {
+		return 0
 	}
 
-	// if the remaining size is smaller than the buffer size, use the remaining size
-	if endSize-currentOffset < chunkSize {
-		chunkSize = endSize - currentOffset
+	// if we didn't read anything yet, assume read the initial offset
+	if totalLogsSize == 0 {
+		totalLogsSize += initialOffset
```
Comment (TomerShor) [inline server/log-collector/pkg/services/logcollector/server.go:1082]: same here, relates to the comment above

**Review by TomerShor — COMMENTED** [review]
Grammar nazi and a question

File: server/log-collector/pkg/services/logcollector/server.go (hunk 2 above)
Comment (liranbg) [reply server/log-collector/pkg/services/logcollector/server.go:376]: u start to read at offset, and you read until size+offset
if you dont add offset, then you start to read at offset and read until size
so basically, for the latter, if you want to read 100 bytes with offset 10, you essentially read 90
see that getChunkSize uses the endsize to calculate the leftToRead in a new way

File: server/log-collector/pkg/services/logcollector/server.go (hunk 3 above)
Comment (liranbg) [reply server/log-collector/pkg/services/logcollector/server.go:1082]: resolves it as the answer relates to https://github.com/mlrun/mlrun/pull/4531#discussion_r1383374543

**Review by quaark — APPROVED** [review]
Good catch!
The only thing I don't like personally is the name of the var `endSize`. Would much rather it be `endIndex`. But I won't die on that hill :)

File: server/log-collector/pkg/services/logcollector/server.go [hunk 4]
```diff
@@ -1055,22 +1066,29 @@ func (s *Server) isLogCollectionRunning(ctx context.Context, runUID, project str
...
 	}
 
-	return chunkSize
+	// if the size we need to read is bigger than the buffer, use the buffer size
+	leftToRead := endSize - totalLogsSize
```
Comment (quaark) [inline server/log-collector/pkg/services/logcollector/server.go:1086]: Actually. Shouldn't this be: ?
```
leftToRead := endSize - totalLogsSize - initialOffset

**Review by quaark — CHANGES_REQUESTED** [review]

**Review by quaark — APPROVED** [review]

**Pull request merged**